pytensor/tensor/optimize.py

```diff
@@ -0,0 +1,523 @@
+import logging
+from collections.abc import Sequence
+from copy import copy
+from typing import cast
+
+import numpy as np
+from scipy.optimize import minimize as scipy_minimize
+from scipy.optimize import minimize_scalar as scipy_minimize_scalar
+from scipy.optimize import root as scipy_root
+
+from pytensor import Variable, function, graph_replace
+from pytensor.gradient import grad, hessian, jacobian
+from pytensor.graph import Apply, Constant, FunctionGraph
+from pytensor.graph.basic import graph_inputs, truncated_graph_inputs
+from pytensor.graph.op import ComputeMapType, HasInnerGraph, Op, StorageMapType
+from pytensor.scalar import bool as scalar_bool
+from pytensor.tensor import dot
+from pytensor.tensor.basic import atleast_2d, concatenate, zeros_like
+from pytensor.tensor.blockwise import Blockwise
+from pytensor.tensor.slinalg import solve
+from pytensor.tensor.variable import TensorVariable
+
+
+_log = logging.getLogger(__name__)
+
+
+class LRUCache1:
+    """
+    Simple LRU cache with a memory size of 1.
+
+    This cache is only usable for a function that takes a single input `x` and returns a single output. The
+    function can also take any number of additional arguments `*args`, but these are assumed to be constant
+    between function calls.
+
+    The purpose of this cache is to allow for Hessian computation to be reused when calling scipy.optimize functions.
+    It is very often the case that some sub-computations are repeated between the objective, gradient, and hessian
+    functions, but by default scipy only allows for the objective and gradient to be fused.
+
+    By using this cache, all 3 functions can be fused, which can significantly speed up the optimization process for
+    expensive functions.
+    """
+
+    def __init__(self, fn):
+        self.fn = fn
+        self.last_x = None
+        self.last_result = None
+
+        self.cache_hits = 0
+        self.cache_misses = 0
+
+        self.value_and_grad_calls = 0
+        self.hess_calls = 0
+
+    def __call__(self, x, *args):
+        """
+        Call the cached function with the given input `x` and additional arguments `*args`.
+
+        If the input `x` is the same as the last input, return the cached result. Otherwise update the cache with the
+        new input and result.
+        """
+        cache_hit = np.all(x == self.last_x)
```

Comment:
Why in this separate variable? You end up doing more checks, you again check for `self.last_x is None`. It's also wasteful here if `self.last_x is None`, you get an array full of Falses, but it still does elementwise comparison between entries and None.

Also miny optimization but doing `(x == self.last_x).all()` is slightly faster, as it goes directly to the C method, instead of the numpy wrapper `np.all` that allows dispatching and all that.

Reply:
I was trying to cover the case where x is a scalar

Reply:
Are you sure? When I run it in a debugger I get `numpy.float64` for `x` in the scalar case

Reply:
Oh I figured it out, it's a dumb scipy thing.